nanofi/examples/tail_file.c

```diff
@@ -0,0 +1,270 @@
+        return;
+    }
+
+    if (get_property(ctx, "delimiter", delimiter, 2) != 0) {
+        return;
+    }
+
+    char delim = '\0';
+    if (strlen(delimiter) > 0) {
+        delim = delimiter[0];
+    }
+
+    if (delim == '\0') {
+        printf("Invalid delimiter \n");
+        return;
+    }
+
+    if (delim == '\\') {
+      if (strlen(delimiter) > 1) {
+        switch (delimiter[1]) {
+          case 'r':
+            delim = '\r';
+            break;
+          case 't':
+            delim = '\t';
+            break;
+          case 'n':
+            delim = '\n';
+            break;
+          case '\\':
+            delim = '\\';
+            break;
+          default:
+            break;
+        }
+      }
+    }
+
+    int curr_offset = get_offset();
+    int max_bytes_read = 4096;
+    char buff[max_bytes_read];
+    FILE * fp = fopen(file_path, "rb");
+    if (!fp) return;
+    fseek(fp, curr_offset, SEEK_SET);
+    while (fgets(buff, max_bytes_read, fp) != NULL) {
+        set_offset(ftell(fp));
+        char * strings[max_bytes_read];
+        int num_strings = 0;
+        for (int i = 0; i < max_bytes_read; ++i) {
+            strings[i] = NULL;
+        }
+        split_string(buff, delim, strings, &num_strings);
+
+        flowfiles = (flow_file_records *)malloc(sizeof(flow_file_records));
+        flowfiles->records = malloc(sizeof(flow_file_record *) * num_strings);
+        flowfiles->len = num_strings;
+
+        for (int i = 0; i < num_strings; ++i) {
+            flowfiles->records[i] = NULL;
+        }
+
+        for (int i = 0; i < num_strings; ++i) {
+            flow_file_record * ffr = generate_flow_file(instance);
+            const char * flow_file = ffr->contentLocation;
+            FILE * ffp = fopen(flow_file, "wb");
+            if (!ffp) {
+                printf("Cannot open flow file at path %s to write content to.\n", flow_file);
+                fclose(fp);
+                free_all_strings(strings, num_strings);
+                return;
+            }
+            fwrite(strings[i], 1, strlen(strings[i]), ffp);
+            flowfiles->records[i] = ffr;
+            free(strings[i]);
+            fclose(ffp);
+        }
+    }
+    fclose(fp);
+}
+
+int main(int argc, char** argv) {
+
+    if (argc < 6) {
+        printf("Error: must run ./tail_file <file> <interval> <delimiter> <nifi instance url> <remote port>\n");
+        exit(1);
+    }
+
+    char * file = argv[1];
+    char * interval = argv[2];
+    char * delimiter = argv[3];
+    char * instance_str = argv[4];
+    char * port_str = argv[5];
+
+    if (access(file, F_OK) == -1) {
+        printf("Error: %s doesn't exist!\n", file);
+        exit(1);
+    }
+
+    struct stat stats;
+    stat(file, &stats);
+
+    // Check for file existence
+    if (S_ISDIR(stats.st_mode)){
+        printf("Error: %s is a directory!\n", file);
+        exit(1);
+    }
+
+    unsigned long intrvl = strtol(interval, NULL, 10);
```

Comment:
According to the documentation of strtol, because all of its return values are potentially valid, you have to set errno to 0 before using it, so that you can differentiate errors from valid cases.
Not zeroing errno and using it afterwards can have unexpected consequences (using a potentially uninitialized value or using an errno set by a previous function).

Reply:
Updated code.